xdsl/transforms/convert_memref_to_ptr.py

```diff
@@ -0,0 +1,164 @@
+
+    match memref_type.layout:
+        case builtin.NoneAttr():
+            strides = builtin.ShapedType.strides_for_shape(memref_type.get_shape())
+        case builtin.StridedLayoutAttr():
+            strides = memref_type.layout.get_strides()
+        case _:
+            raise DiagnosticException(f"Unsupported layout type {memref_type.layout}")
+
+    ops: list[Operation] = []
+
+    head: SSAValue | None = None
+
+    for index, stride in zip(indices, strides, strict=True):
+        # Calculate the offset that needs to be added through the index of the current
+        # dimension.
+        increment = index
+        match stride:
+            case None:
+                raise DiagnosticException(
+                    f"MemRef {memref_type} with dynamic stride is not yet implemented"
+                )
+            case 1:
+                # Stride 1 is a noop making the index equal to the offset.
+                pass
+            case _:
+                # Otherwise, multiply the stride (which by definition is the number of
+                # elements required to be skipped when incrementing that dimension).
+                ops.extend(
+                    (
+                        stride_op := arith.Constant.from_int_and_width(
+                            stride, builtin.IndexType()
+                        ),
+                        offset_op := arith.Muli(increment, stride_op),
+                    )
+                )
+                stride_op.result.name_hint = "pointer_dim_stride"
+                offset_op.result.name_hint = "pointer_dim_offset"
+
+                increment = offset_op.result
+
+        if head is None:
+            # First iteration.
+            head = increment
+            continue
+
+        # Otherwise sum up the products.
+        add_op = arith.Addi(head, increment)
+        add_op.result.name_hint = "pointer_dim_stride"
+        ops.append(add_op)
+        head = add_op.result
+
+    if head is None:
+        raise DiagnosticException("Got empty indices for offset calculations.")
+
+    ops.extend(
+        [
+            bytes_per_element_op := ptr.TypeOffsetOp(
+                operands=[],
+                result_types=[builtin.IndexType()],
+                properties={"elem_type": memref_type.element_type},
+            ),
+            final_offset := arith.Muli(head, bytes_per_element_op),
+        ]
+    )
+
+    bytes_per_element_op.offset.name_hint = "bytes_per_element"
+    final_offset.result.name_hint = "scaled_pointer_offset"
+
+    return ops, final_offset.result
+
+
+def get_target_ptr(
+    target_memref: SSAValue,
+    memref_type: memref.MemRefType[Any],
+    indices: Iterable[SSAValue],
+) -> tuple[list[Operation], SSAValue]:
+    ops: list[Operation] = [
+        memref_ptr := ptr.ToPtrOp(
+            operands=[target_memref], result_types=[ptr.PtrType()]
+        )
+    ]
+
+    if len(list(indices)) == 0:
```

Comment:
```suggestion
    if not indices:
```

Reply:
It looks like the indices passed below are actually a Collection, so this should be possible, and would avoid an extra list construction